### PR TITLE
bisq: init at 1.7.0

### LINKED ISSUE #69807: Add bisq
**Project description**
It's a p2p marketplace for crypto currencies.

**Metadata**

* homepage URL: https://bisq.network/
* source URL: https://github.com/bisq-network/bisq
* license: AGPL 3
* platforms: linux, mac

### PR DESCRIPTION
**Pull request by emmanuelrosa**
###### Motivation for this change

This change adds the Bisq decentralized bitcoin exchange.

Closes #69807

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/blockchains/bisq-desktop/grpc-java.nix [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  nativeBuildInputs = [
+    autoPatchelfHook
+  ];
+
+  dontUnpack = true;
+
+  installPhase = ''
+    install -m755 -D $src $out/bin/protoc-gen-rpc-java
+  '';
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/grpc-java.nix:19]: Missing meta.

File: pkgs/applications/blockchains/bisq-desktop/mkdeps.nix [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+with import <nixpkgs> { };
+let
+  common = callPackage ./common.nix {};
+  jdk = common.jdk;
+  version = common.version;
+  pname = common.pname;
+  src = common.src;
+  grpc = common.grpc;
+  gradle = common.gradle;
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/mkdeps.nix:26]:
```suggestion
  inherit (common) jdk version pname grpc gradle;
```

File: pkgs/applications/blockchains/bisq-desktop/mkdeps.nix [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    # perl code mavenizes pathes (com.squareup.okio/okio/1.13.0/a9283170b7305c8d92d25aff02a6ab7e45d06cbe/okio-1.13.0.jar -> com/squareup/okio/okio/1.13.0/okio-1.13.0.jar)
+    installPhase = ''
+      find $GRADLE_USER_HOME -type f -regex '.*\.\(jar\|pom\)' \
+        | perl -pe 's#(.*/([^/]+)/([^/]+)/([^/]+)/[0-9a-f]{30,40}/([^/\s]+))$# ($x = $2) =~ tr|\.|/|; "install -Dm444 $1 \$out/$x/$3/$4/$5" #e' \
+        | sh
+
+   '';
+
+    dontStrip = true;
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/mkdeps.nix:54]: Why are we not stripping?

File: pkgs/applications/blockchains/bisq-desktop/default.nix [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    )
+  '';
+
+  installPhase = ''
+    mkdir -p $out/lib
+    cp -r lib/* $out/lib
+    for jar in $out/lib/*.jar; do
+      classpath="$classpath:$jar"
+    done
+    mkdir -p $out/bin
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/default.nix:79]:
```suggestion
```

File: pkgs/applications/blockchains/bisq-desktop/default.nix [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        --replace 'jcenter()' 'mavenLocal(); maven { url uri("${deps}") }' \
+        --replace 'artifact = "com.google.protobuf:protoc:''${protocVersion}"' "path = '${protobuf3_10}/bin/protoc'" \
+        --replace 'artifact = "io.grpc:protoc-gen-grpc-java:''${grpcVersion}"' "path = '${grpc}/bin/protoc-gen-rpc-java'"
+      sed -i 's/^.*com.github.JesusMcCloud.tor-binary:tor-binary-linux64.*$//g' gradle/witness/gradle-witness.gradle
+      gradle  --offline --no-daemon desktop:build  --exclude-task desktop:test
+    )
+  '';
+
+  installPhase = ''
+    mkdir -p $out/lib
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/default.nix:74]:
```suggestion
    mkdir -p $out/lib $out/bin
```

File: pkgs/applications/blockchains/bisq-desktop/default.nix [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      classpath="$classpath:$jar"
+    done
+    mkdir -p $out/bin
+    makeWrapper ${jdk}/bin/java $out/bin/bisq-desktop-wrapped \
+      --add-flags "-classpath $classpath bisq.desktop.app.BisqAppMain"
+    makeWrapper ${bisq-launcher} $out/bin/bisq-desktop \
+      --prefix PATH : $out/bin
+
+    install -Dm644 -t $out/share/applications \
+      ${desktopItem}/share/applications/*
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/default.nix:86]: Please use copyDesktopItems.

File: pkgs/applications/blockchains/bisq-desktop/default.nix [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    for n in 16 24 32 48 64 96 128 256; do
+      size=$n"x"$n
+      ${imagemagick}/bin/convert $src/desktop/package/linux/icon.png -resize $size bisq.png
+      install -Dm644 \
+        -t $out/share/icons/hicolor/$size/apps \
+        bisq.png
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/default.nix:93]:
```suggestion
      install -Dm644 -t $out/share/icons/hicolor/$size/apps bisq.png
```

File: pkgs/applications/blockchains/bisq-desktop/default.nix [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        for p in ${toString d}; do
+          ${rsync}/bin/rsync -a $p/ $out/
+        done
+      '';
+    };
+in stdenv.mkDerivation rec {
+  inherit (common) pname version src jdk grpc gradle;
+
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ gradle ];
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/default.nix:48]: If gradle is executed on the build machine it needs to go to nativeBuildInputs.

File: pkgs/applications/blockchains/bisq-desktop/default.nix [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    d = map mkDepDerivation (import ./deps.nix); 
+    in stdenv.mkDerivation {
+      name = "bisq-deps";
+      dontUnpack = true;
+
+      installPhase = ''
+        mkdir $out
+
+        for p in ${toString d}; do
+          ${rsync}/bin/rsync -a $p/ $out/
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/default.nix:40]: Why are we not using cp here?

File: pkgs/applications/blockchains/bisq-desktop/common.nix [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  }).overrideAttrs (oldAttrs: {
+    nativeBuildInputs = oldAttrs.nativeBuildInputs or [] ++ [ git-lfs ];
+  });
+
+  # We use a modified JDK package due to an issue with libcrypto.
+  # See https://github.com/NixOS/nixpkgs/issues/73246
+  jdk = openjdk11.overrideAttrs (oldAttrs: rec {
+    buildInputs = lib.remove gnome2.gnome_vfs oldAttrs.buildInputs;
+    NIX_LDFLAGS = builtins.replaceStrings [ "-lgnomevfs-2" ] [ "" ] oldAttrs.NIX_LDFLAGS;
+  });
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/common.nix:34]: I think https://github.com/NixOS/nixpkgs/pull/73407 would be a cleaner solution which should be chosen instead.

File: pkgs/applications/blockchains/bisq-desktop/common.nix [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    sha256 = "1zmf76i4yddr4zc2jcm09bgs7yya6bqv1zk68z17g3r39qmyxv1q";
+    postFetch = ''
+      cd $out
+      git clone $url
+      cd bisq
+      git lfs install --force --local
+      git lfs pull
+      cp -v p2p/src/main/resources/* $out/p2p/src/main/resources/
+      cd ..
+      rm -rf bisq
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/common.nix:23]:
```suggestion
      rm -r bisq
```

**Comment by SuperSandro2000** [comment]
Also please remove the trailing whitespace the editorCheck found.

File: pkgs/applications/blockchains/bisq-desktop/default.nix (hunk 9 above)
Comment (emmanuelrosa) [reply pkgs/applications/blockchains/bisq-desktop/default.nix:40]: I tried `cp` but it didn't work.

File: pkgs/applications/blockchains/bisq-desktop/common.nix (hunk 10 above)
Comment (emmanuelrosa) [reply pkgs/applications/blockchains/bisq-desktop/common.nix:34]: The solution implemented in this PR was written by the author of #73407

File: pkgs/applications/blockchains/bisq-desktop/default.nix (hunk 9 above)
Comment (emmanuelrosa) [reply pkgs/applications/blockchains/bisq-desktop/default.nix:40]: To be more specific, I could not get `cp` to create a proper Maven directory tree.

Comment (SuperSandro2000) [reply pkgs/applications/blockchains/bisq-desktop/default.nix:48]: I need more details if I should help you but I am pretty sure cp will work here but you maybe need to try with ``./`` or ``/`` a bit.

File: pkgs/applications/blockchains/bisq-desktop/common.nix (hunk 10 above)
Comment (SuperSandro2000) [reply pkgs/applications/blockchains/bisq-desktop/common.nix:40]: Yeah but this should be probably done globally do not build a different openjdk for this package.

**Comment by emmanuelrosa** [comment]
I'm going to revert using `copyDesktopItems` and just copy it manually because I can't get it to work.

File: pkgs/applications/blockchains/bisq-desktop/mkdeps.nix [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+  prebuild = pkgs.stdenv.mkDerivation {
+    name = "${pname}-prebuild-${version}";
+    inherit src;
+    buildInputs = [ gradle pkgs.perl pkgs.unzip pkgs.zip ];
+
+    patchPhase = ''
+      substituteInPlace ./build.gradle \
+        --replace 'artifact = "com.google.protobuf:protoc:''${protocVersion}"' "path = '${pkgs.protobuf3_10}/bin/protoc'"
+      substituteInPlace ./build.gradle \
+        --replace 'artifact = "io.grpc:protoc-gen-grpc-java:''${grpcVersion}"' "path = '${grpc}/bin/protoc-gen-rpc-java'"
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/mkdeps.nix:33]:
```suggestion
      substituteInPlace ./build.gradle \
        --replace 'artifact = "com.google.protobuf:protoc:''${protocVersion}"' "path = '${pkgs.protobuf3_10}/bin/protoc'" \
        --replace 'artifact = "io.grpc:protoc-gen-grpc-java:''${grpcVersion}"' "path = '${grpc}/bin/protoc-gen-rpc-java'"
```

File: pkgs/applications/blockchains/bisq-desktop/mkdeps.nix [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+      gradle --no-daemon desktop:build --exclude-task desktop:test
+    '';
+
+    # perl code mavenizes pathes (com.squareup.okio/okio/1.13.0/a9283170b7305c8d92d25aff02a6ab7e45d06cbe/okio-1.13.0.jar -> com/squareup/okio/okio/1.13.0/okio-1.13.0.jar)
+    installPhase = ''
+      find $GRADLE_USER_HOME -type f -regex '.*\.\(jar\|pom\)' \
+        | perl -pe 's#(.*/([^/]+)/([^/]+)/([^/]+)/[0-9a-f]{30,40}/([^/\s]+))$# ($x = $2) =~ tr|\.|/|; "install -Dm444 $1 \$out/$x/$3/$4/$5" #e' \
+        | sh
+
+    '';
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/mkdeps.nix:47]:
```suggestion
        | sh
    '';
```

**Comment by SuperSandro2000** [comment]
> I'm going to revert using `copyDesktopItems` and just copy it manually because I can't get it to work.

Did you use it like https://github.com/NixOS/nixpkgs/blob/7248f89da2dab8bca3a9f4f11d0407d0c8e25e89/pkgs/applications/misc/prusa-slicer/default.nix#L89 ?

File: pkgs/applications/blockchains/bisq-desktop/common.nix (hunk 10 above)
Comment (juaningan) [reply pkgs/applications/blockchains/bisq-desktop/common.nix:40]: Hi, this was a workaround for a jre crash when Bisq ask for wallet password. Bug is not there anymore in last nixpkgs commits. So you can go with vanilla `openjdk11`.

**Comment by emmanuelrosa** [comment]
Yes, I used `copyDesktopItems` as in the examples. The hook was not triggering. It works if I call `copyDesktopItems` in the install script, which suggests the `postInstallHooks` part is not working.

**Comment by jb55** [comment]
I noticed an insane amount of resident memory usage (5.6 GB) I mentioned this on twitter and the bisq folks suggested:

https://twitter.com/bisq_network/status/1385729925563355138

`JAVA_TOOL_OPTIONS="-XX:MaxRAM=4g" bisq-desktop`

maybe something to consider adding to the launch script?

**Comment by emmanuelrosa** [comment]
Does Bisq continue to function properly when constrained to 4G of RAM? In a nutshell, is it still possible to trade? Are there any side-effects?

**Comment by jb55** [comment]
On Fri, Apr 23, 2021 at 04:27:15PM -0700, Emmanuel Rosa wrote:
>Does Bisq continue to function properly when constrained to 4G of RAM?
>In a nutshell, is it still possible to trade? Are there any
>side-effects?

I haven't noticed any issues but if we want to be conservative we don't have to worry about it right now. Perhaps we can look into it if we get more complaints about it. At least we've documented a workaround here for now.

File: pkgs/applications/blockchains/bisq-desktop/common.nix [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+, openjdk11
+, gradle
+, gradleGen
+, gnome2
+}:
+rec {
+  pname = "bisq-desktop";
+  version = "1.6.2";
+
+  src = (fetchgit rec {
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/common.nix:17]: Please use fetchFromGitHub.

File: pkgs/applications/blockchains/bisq-desktop/mkdeps.nix [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+    echo "]"
+  '';
+
+in
+pkgs.stdenv.mkDerivation {
+  name = "${pname}-deps-${version}.nix";
+
+  phases = [ "installPhase" ];
+
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/mkdeps.nix:99]:
```suggestion
```
Please do not overwrite the phases and disable the ones which are not required.

File: pkgs/applications/blockchains/bisq-desktop/mkdeps.nix [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+      echo "    mavenDir = \"$upstreamDir\";"
+      echo "  }"
+    done
+
+    echo "]"
+  '';
+
+in
+pkgs.stdenv.mkDerivation {
+  name = "${pname}-deps-${version}.nix";
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/mkdeps.nix:96]: If we hardcode the version the deps need to be rebuild for every update even if they stay the same.

File: pkgs/applications/blockchains/bisq-desktop/grpc-java.nix [hunk 17]
```diff
@@ -0,0 +1,28 @@
+{ stdenv, lib, fetchurl, autoPatchelfHook }:
+
+stdenv.mkDerivation rec {
+  name = "protoc-gen-grpc-java";
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/bisq-desktop/grpc-java.nix:4]:
```suggestion
  pname = "protoc-gen-grpc-java";
```

**Comment by emmanuelrosa** [comment]
When using `fetchFromGithub`, does the option `fetchLFS` affect the SHA256 of the derivation? If so, that's going to be a problem because the content pulled using Git LFS can change independently from the source code. I'm facing a hash mismatch out of the blue, and I suspect it's because the P2P data that's pulled in using Git LFS has changed.

**Comment by emmanuelrosa** [comment]
OK the problem is as follows.

Bisq distributions come with snapshots of its various databases. These files are included in the p2p.jar file. The files were once included in the Git repo, but as they continued to grow in size, they were later excluded and replaced with providing the files using Git LFS. There's a documented process for producing the files using a Bisq client that's synchronized. Importantly, Bisq will not run without these files.

This creates a problem for this package because there's no place to download these files in a reproducible way; Actually, using a hack there is a way. Below are a few options I've been pondering:

 1. Have the package maintainers provide online reproducible storage for the data files. Obviously this is my least favorite idea.
 2. Move the responsibility to users by having the launcher obtain the data files (somehow) and inserting them into p2p.jar, prior to launching Bisq. Or, have a script which users can run at their leisure, but this can have unintended consequences, such as the data becoming older than what the client expects. Either option would impact the UX. 
 3. Use the upstream .dep distribution instead of compiling from source.
 4. A hybrid; Continue to build from source, but use the corresponding .dep distribution as a source for the data files by extracting them from p2p.jar. This would effectively use an existing reproducible store to obtain the data files. It's a hack, sure, but it should work.

 Which option do you think is best?

**Comment by emmanuelrosa** [comment]
I refactored the package using upstream's .deb. It would mean not building from source, but it's the most maintainable option. I'm only using the Java code and not any of the native libraries provided by upstream, since they would need a lot of patchelf-ing, and thus far it's working fine without them. Most likely the JRE is taking care of it. I going to test it some more before pushing it here.

**Comment by cyplo** [comment]
works very well for me :)

**Comment by emmanuelrosa** [comment]
Note: Bisq-1.7.0 will be the mandatory minimum version in about a week from now, due to a privacy issue that has been fixed. I've already updated my NUR, but it may take an hour or so to show up in the NixOS nur repo.

**Pull request merged**